diff -urN samba-2.0.7.org/source/Makefile.in samba-2.0.7/source/Makefile.in
--- samba-2.0.7.org/source/Makefile.in	Thu Oct 19 00:01:25 2000
+++ samba-2.0.7/source/Makefile.in	Fri Oct 20 22:58:16 2000
@@ -21,15 +21,16 @@
 builddir=@builddir@
 SHELL=/bin/sh
 
-BASEDIR= @prefix@
-BINDIR = @bindir@
+BASEDIR= $(DESTDIR)@prefix@
+BINDIR = $(DESTDIR)@bindir@
 # we don't use sbindir because we want full compatibility with
 # the previous releases of Samba
-SBINDIR = @bindir@
-LIBDIR = @libdir@
-VARDIR = @localstatedir@
-MANDIR = @mandir@
-SAMBABOOK = @sambabook@
+SBINDIR = $(DESTDIR)@sbindir@
+LIBDIR = $(DESTDIR)@libdir@
+VARDIR = $(DESTDIR)@localstatedir@
+MANDIR = $(DESTDIR)@mandir@
+SYSCONFDIR = $(DESTDIR)@sysconfdir@
+SAMBABOOK = $(DESTDIR)@sambabook@
 
 # The permissions to give the executables
 INSTALLPERMS = 0755
@@ -49,7 +50,7 @@
 SMB_PASSWD_FILE = $(PRIVATEDIR)/smbpasswd
 
 # This is where SWAT images and help files go
-SWATDIR = @swatdir@
+SWATDIR = $(DESTDIR)@swatdir@
 
 # the directory where lock files go
 LOCKDIR = @lockdir@
diff -urN samba-2.0.7.org/source/script/installbin.sh samba-2.0.7/source/script/installbin.sh
--- samba-2.0.7.org/source/script/installbin.sh	Thu Oct 19 00:01:26 2000
+++ samba-2.0.7/source/script/installbin.sh	Fri Oct 20 23:12:27 2000
@@ -13,7 +13,7 @@
 
 for d in $BASEDIR $BINDIR $LIBDIR $VARDIR; do
 if [ ! -d $d ]; then
-mkdir $d
+mkdir -p $d
 if [ ! -d $d ]; then
   echo Failed to make directory $d
   exit 1
@@ -34,7 +34,7 @@
 
  # this is a special case, mount needs this in a specific location
  if [ $p2 = smbmount ]; then
-   ln -sf $BINDIR/$p2 /sbin/mount.smbfs
+   ln -sf /usr/sbin/$p2 $DESTDIR/$SBINDIR/mount.smbfs
  fi
 done
 
diff -urN samba-2.0.7.org/source/script/installcp.sh samba-2.0.7/source/script/installcp.sh
--- samba-2.0.7.org/source/script/installcp.sh	Thu Oct 19 00:01:26 2000
+++ samba-2.0.7/source/script/installcp.sh	Fri Oct 20 23:07:24 2000
@@ -12,7 +12,7 @@
 echo Installing codepage files in $CODEPAGEDIR
 for d in $LIBDIR $CODEPAGEDIR; do
 if [ ! -d $d ]; then
-mkdir $d
+mkdir -p $d
 if [ ! -d $d ]; then
   echo Failed to make directory $d
   exit 1
@@ -23,11 +23,11 @@
 for p in $*; do
  if [ -f ${srcdir}/codepages/codepage_def.$p ]; then
    echo Creating codepage file $CODEPAGEDIR/codepage.$p
-   $BINDIR/make_smbcodepage c $p ${srcdir}/codepages/codepage_def.$p $CODEPAGEDIR/codepage.$p
+   ${srcdir}/bin/make_smbcodepage c $p ${srcdir}/codepages/codepage_def.$p $CODEPAGEDIR/codepage.$p
  fi
  if [ -f ${srcdir}/codepages/CP${p}.TXT ]; then
    echo Creating unicode map $CODEPAGEDIR/unicode_map.$p
-   $BINDIR/make_unicodemap $p ${srcdir}/codepages/CP${p}.TXT $CODEPAGEDIR/unicode_map.$p
+   ${srcdir}/bin/make_unicodemap $p ${srcdir}/codepages/CP${p}.TXT $CODEPAGEDIR/unicode_map.$p
  fi
 done
 
diff -urN samba-2.0.7.org/source/script/installman.sh samba-2.0.7/source/script/installman.sh
--- samba-2.0.7.org/source/script/installman.sh	Thu Oct 19 00:01:26 2000
+++ samba-2.0.7/source/script/installman.sh	Fri Oct 20 22:57:52 2000
@@ -11,7 +11,7 @@
 
 for d in $MANDIR $MANDIR/man1 $MANDIR/man5 $MANDIR/man7 $MANDIR/man8; do
 if [ ! -d $d ]; then
-mkdir $d
+mkdir -p $d
 if [ ! -d $d ]; then
   echo Failed to make directory $d, does $USER have privileges?
   exit 1
diff -urN samba-2.0.7.org/source/script/installscripts.sh samba-2.0.7/source/script/installscripts.sh
--- samba-2.0.7.org/source/script/installscripts.sh	Thu Oct 19 00:01:26 2000
+++ samba-2.0.7/source/script/installscripts.sh	Fri Oct 20 22:57:52 2000
@@ -12,7 +12,7 @@
 
 for d in $BINDIR; do
  if [ ! -d $d ]; then
-  mkdir $d
+  mkdir -p $d
   if [ ! -d $d ]; then
     echo Failed to make directory $d
     echo Have you run installbin first?
diff -urN samba-2.0.7.org/source/script/installswat.sh samba-2.0.7/source/script/installswat.sh
--- samba-2.0.7.org/source/script/installswat.sh	Thu Oct 19 00:01:26 2000
+++ samba-2.0.7/source/script/installswat.sh	Fri Oct 20 23:00:27 2000
@@ -10,7 +10,7 @@
 
 for d in $SWATDIR $SWATDIR/help $SWATDIR/images $SWATDIR/include; do
     if [ ! -d $d ]; then
-        mkdir $d
+        mkdir -p $d
         if [ ! -d $d ]; then
             echo Failed to make directory $d, does $USER have privileges?
             exit 1
@@ -69,7 +69,7 @@
 
     for d in $BOOKDIR $BOOKDIR/figs $BOOKDIR/gifs; do
         if [ ! -d $d ]; then
-            mkdir $d
+            mkdir -p $d
             if [ ! -d $d ]; then
                 echo Failed to make directory $d, does $USER have privileges?
                 exit 1
