diff -Nur samba-2.0.0/source/include/includes.h samba-2.0.0.pld/source/include/includes.h
--- samba-2.0.0/source/include/includes.h	Tue Jan 26 18:43:00 1999
+++ samba-2.0.0.pld/source/include/includes.h	Tue Jan 26 18:42:43 1999
@@ -310,7 +310,7 @@
 #endif
 
 #ifdef HAVE_SYS_CAPABILITY_H
-#include <sys/capability.h>
+ /* #include <sys/capability.h> .. not works with glibc-2.0.109 ;( */
 #endif
 
 #if defined(HAVE_RPC_RPC_H)
