--- samba-3.0.25a/source/Makefile.in.orig	2007-06-16 18:09:12.204259000 +0200
+++ samba-3.0.25a/source/Makefile.in	2007-06-16 18:13:52.122097878 +0200
@@ -1174,7 +1174,7 @@
 
 bin/libmsrpc.@SHLIBEXT@: proto_exists $(CAC_OBJ)
 	@echo Linking libmsrpc shared library $@
-	@$(SHLD) $(LDSHFLAGS) -o $@ $(CAC_OBJ) $(LIBS) \
+	@$(SHLD) $(LDSHFLAGS) -o $@ $(CAC_OBJ) $(LIBS) $(LDAP_LIBS) $(KRB5LIBS) $(NSCD_LIBS) \
 	@SONAMEFLAG@`basename $@`.$(LIBMSRPC_MAJOR)
 
 bin/libmsrpc.a: proto_exists $(CAC_OBJ)
