diff -urN samba-2.2.1.org/source/Makefile.in samba-2.2.1/source/Makefile.in
--- samba-2.2.1.org/source/Makefile.in	Sun Jul  8 22:29:34 2001
+++ samba-2.2.1/source/Makefile.in	Wed Jul 11 19:56:04 2001
@@ -49,7 +49,7 @@
 CONFIGFILE = $(CONFIGDIR)/smb.conf
 LMHOSTSFILE = $(CONFIGDIR)/lmhosts
 DRIVERFILE = $(CONFIGDIR)/printers.def
-PASSWD_PROGRAM = /bin/passwd
+PASSWD_PROGRAM = /usr/bin/passwd
 # This is where smbpasswd et al go
 PRIVATEDIR = @privatedir@
 
diff -urN samba-2.2.1.org/source/configure.in samba-2.2.1/source/configure.in
--- samba-2.2.1.org/source/configure.in	Tue Jul 10 02:12:50 2001
+++ samba-2.2.1/source/configure.in	Wed Jul 11 20:00:03 2001
@@ -656,6 +656,11 @@
 		AC_DEFINE(HAVE_STRCASECMP))
 fi
 
+# Check for yp_get_default_domain in -lnsl (linux glibc2.1)
+#
+AC_LIBTESTFUNC(nsl, yp_get_default_domain)
+#
+
 #
 # Check for the functions putprpwnam, set_auth_parameters,
 # getspnam, bigcrypt and getprpwnam in -lsec and -lsecurity
@@ -1630,7 +1635,7 @@
   yes)
     AC_MSG_RESULT(yes)
     AC_DEFINE(WITH_PAM)
-    LIBS="$LIBS -lpam"
+    LIBS="$LIBS -ldl -lpam -lpamc"
     with_pam_for_crypt=yes
     ;;
   *)
@@ -1776,20 +1781,18 @@
   yes)
     AC_MSG_RESULT(yes)
     AC_DEFINE(WITH_SSL)
-    withval="/usr/local/ssl"     # default
-
+    withval="/usr"     # default
     if test "${with_sslinc+set}" = set; then
 
 	withval="$with_sslinc"
 	case "$withval" in
 	yes|no)
 	  echo "configure: warning: --with-sslinc called without argument - will use default" 1>&w
-	  CFLAGS="-I/usr/local/ssl/include $CFLAGS"
+	 CFLAGS="-I/usr/include/openssl $CFLAGS"
 	  LIBS="-lssl -lcrypto $LIBS"
-	  LDFLAGS="=L/usr/local/ssl/lib $LDFLAGS"
 	;;
         * )
-          CFLAGS="-I${withval} $CFLAGS"
+	  CFLAGS="-I${withval}/include/openssl $CFLAGS"
 	  LIBS="-lssl -lcrypto $LIBS"
   	  LDFLAGS="-L${withval}/lib $LDFLAGS"
 	;;
diff -urN samba-2.2.1.org/source/smbwrapper/smbsh.in samba-2.2.1/source/smbwrapper/smbsh.in
--- samba-2.2.1.org/source/smbwrapper/smbsh.in	Wed Oct  7 06:43:09 1998
+++ samba-2.2.1/source/smbwrapper/smbsh.in	Wed Jul 11 20:00:35 2001
@@ -1,6 +1,6 @@
 #! /bin/sh
 
-SMBW_LIBDIR=${SMBW_LIBDIR-@builddir@/smbwrapper}
+SMBW_LIBDIR=${SMBW_LIBDIR-/lib/security}
 
 if [ ! -f ${SMBW_LIBDIR}/smbwrapper.so ]; then
     echo You need to set LIBDIR in smbsh
