diff -urN samba-3.0.4.org/source/aclocal.m4 samba-3.0.4/source/aclocal.m4
--- samba-3.0.4.org/source/aclocal.m4	2004-05-13 02:29:15.224209664 +0200
+++ samba-3.0.4/source/aclocal.m4	2004-05-13 02:29:22.478106904 +0200
@@ -541,12 +541,12 @@
   $1=[$]ac_new_flags
 ])
     
-dnl Removes -L/usr/lib/? from given variable
+dnl Removes -L/usr/lib64/? from given variable
 AC_DEFUN(LIB_REMOVE_USR_LIB,[
   ac_new_flags=""
   for i in [$]$1; do
     case [$]i in
-    -L/usr/lib|-L/usr/lib/) ;;
+    -L/usr/lib64|-L/usr/lib64/) ;;
     *) ac_new_flags="[$]ac_new_flags [$]i" ;;
     esac
   done
@@ -677,9 +677,9 @@
 ])
 
 dnl LIB_ADD_DIR(LDFLAGS, $LIBDIR)
-dnl This function doesn't add -L/usr/lib into LDFLAGS
+dnl This function doesn't add -L/usr/lib64 into LDFLAGS
 AC_DEFUN(LIB_ADD_DIR,[
-if test "$2" != "/usr/lib" ; then
+if test "$2" != "/usr/lib64" ; then
     $1="$$1 -L$2"
 fi
 ])
diff -urN samba-3.0.4.org/source/configure.in samba-3.0.4/source/configure.in
--- samba-3.0.4.org/source/configure.in	2004-05-13 02:29:15.234208144 +0200
+++ samba-3.0.4/source/configure.in	2004-05-13 02:32:02.287812136 +0200
@@ -19,7 +19,7 @@
 mandir="\${prefix}/man"
 logfilebase="\${VARDIR}"
 privatedir="\${prefix}/private"
-libdir="\${prefix}/lib"
+libdir="\${prefix}/lib64"
 configdir="\${LIBDIR}"
 swatdir="\${prefix}/swat"
 
@@ -27,12 +27,12 @@
 [  --with-fhs              Use FHS-compliant paths (default=no)],
 [ case "$withval" in
   yes)
-    lockdir="\${VARDIR}/lib/samba"
+    lockdir="\${VARDIR}/lib64/samba"
     piddir="\${VARDIR}/run"
     mandir="\${prefix}/share/man"
     logfilebase="\${VARDIR}/log/samba"
     privatedir="\${CONFIGDIR}/private"
-    libdir="\${prefix}/lib/samba"
+    libdir="\${prefix}/lib64/samba"
     configdir="${sysconfdir}/samba"
     swatdir="\${DATADIR}/samba/swat"
     ;;
@@ -587,7 +587,7 @@
 # Note: iconv does that explicitly below, but other packages
 # don't.
 	CPPFLAGS="$CPPFLAGS -I/sw/include"
-	LDFLAGS="$LDFLAGS -L/sw/lib"
+	LDFLAGS="$LDFLAGS -L/sw/lib64"
 
 # If we have dlsym_prepend_underscore (from Fink's dlcompat),
 # use that instead of plain dlsym.
@@ -868,7 +868,7 @@
     _cppflags=${CPPFLAGS}
 
     # Add additional search path
-    LDFLAGS="-L$with_readline/lib $LDFLAGS"
+    LDFLAGS="-L$with_readline/lib64 $LDFLAGS"
     CPPFLAGS="-I$with_readline/include $CPPFLAGS"
 
     AC_CHECK_HEADERS(readline.h history.h readline/readline.h)
@@ -879,7 +879,7 @@
        AC_CHECK_LIB(${termlib}, tgetent, [TERMLIBS="-l${termlib}"; break])
       done
       AC_CHECK_LIB(readline, rl_callback_handler_install,
-       [TERMLDFLAGS="-L$with_readline/lib"
+       [TERMLDFLAGS="-L$with_readline/lib64"
        TERMCPPFLAGS="-I$with_readline/include"
        CPPFLAGS="-I$with_readline/include $CPPFLAGS"
        TERMLIBS="-lreadline $TERMLIBS"
@@ -1658,7 +1658,7 @@
 ICONV_LOCATION=standard
 LOOK_DIRS="/usr /usr/local /sw /opt"
 AC_ARG_WITH(libiconv,
-[  --with-libiconv=BASEDIR Use libiconv in BASEDIR/lib and BASEDIR/include (default=auto) ],
+[  --with-libiconv=BASEDIR Use libiconv in BASEDIR/lib64 and BASEDIR/include (default=auto) ],
 [
   if test "$withval" = "no" ; then
     AC_MSG_ERROR([argument to --with-libiconv must be a directory])
@@ -1708,7 +1708,7 @@
         # check for iconv in libc
         ic_save_LIBS="$LIBS"
         if test x"$ICONV_PATH_SPEC" = "xyes" ; then
-           LIBS="$LIBS -L$ICONV_LOCATION/lib"
+           LIBS="$LIBS -L$ICONV_LOCATION/lib64"
         fi
         if test x"$jm_cv_lib_iconv" != x; then
            LIBS="$LIBS -l$jm_cv_lib_iconv"
@@ -2568,7 +2568,7 @@
         AC_MSG_RESULT($withval)
         KRB5_CFLAGS="-I$withval/include"
         KRB5_CPPFLAGS="-I$withval/include"
-        KRB5_LDFLAGS="-L$withval/lib"
+        KRB5_LDFLAGS="-L$withval/lib64"
         FOUND_KRB5=yes
         ;;
       esac ],
@@ -2581,10 +2581,10 @@
     # see if this box has the SuSE location for the heimdal krb implementation
     AC_MSG_CHECKING(for /usr/include/heimdal)
     if test -d /usr/include/heimdal; then
-      if test -f /usr/lib/heimdal/lib/libkrb5.a; then
+      if test -f /usr/lib64/heimdal/lib/libkrb5.a; then
           KRB5_CFLAGS="-I/usr/include/heimdal"
           KRB5_CPPFLAGS="-I/usr/include/heimdal"
-          KRB5_LDFLAGS="-L/usr/lib/heimdal/lib"
+          KRB5_LDFLAGS="-L/usr/lib64/heimdal/lib"
           AC_MSG_RESULT(yes)
       else
           KRB5_CFLAGS="-I/usr/include/heimdal"
@@ -4372,7 +4372,7 @@
 builddir=`pwd`
 AC_SUBST(builddir)
 
-dnl Remove -L/usr/lib/? from LDFLAGS and LIBS
+dnl Remove -L/usr/lib64/? from LDFLAGS and LIBS
 LIB_REMOVE_USR_LIB(LDFLAGS)
 LIB_REMOVE_USR_LIB(LIBS)
 
diff -urN samba-3.0.4.org/source/Makefile.in samba-3.0.4/source/Makefile.in
--- samba-3.0.4.org/source/Makefile.in	2004-05-13 02:29:15.223209816 +0200
+++ samba-3.0.4/source/Makefile.in	2004-05-13 02:29:22.474107512 +0200
@@ -1278,9 +1278,9 @@
 	@$(SHELL) $(srcdir)/script/installswat.sh $(DESTDIR)$(SWATDIR) $(srcdir)
 
 installclientlib: installdirs libsmbclient
-	@$(SHELL) $(srcdir)/script/installdirs.sh $(DESTDIR)${prefix}/lib
-	-$(INSTALLCLIENTCMD_SH) bin/libsmbclient.@SHLIBEXT@ $(DESTDIR)${prefix}/lib
-	-$(INSTALLCLIENTCMD_A) bin/libsmbclient.a $(DESTDIR)${prefix}/lib
+	@$(SHELL) $(srcdir)/script/installdirs.sh $(DESTDIR)${prefix}/lib64
+	-$(INSTALLCLIENTCMD_SH) bin/libsmbclient.@SHLIBEXT@ $(DESTDIR)${prefix}/lib64
+	-$(INSTALLCLIENTCMD_A) bin/libsmbclient.a $(DESTDIR)${prefix}/lib64
 	@$(SHELL) $(srcdir)/script/installdirs.sh $(DESTDIR)${prefix}/include
 	-$(INSTALLCMD) $(srcdir)/include/libsmbclient.h $(DESTDIR)${prefix}/include
 
diff -urN samba-3.0.4.org/source/torture/nsstest.c samba-3.0.4/source/torture/nsstest.c
--- samba-3.0.4.org/source/torture/nsstest.c	2004-05-13 02:29:15.062234288 +0200
+++ samba-3.0.4/source/torture/nsstest.c	2004-05-13 02:29:22.503103104 +0200
@@ -21,7 +21,7 @@
 
 #include "includes.h"
 
-static const char *so_path = "/lib/libnss_winbind.so";
+static const char *so_path = "/lib64/libnss_winbind.so";
 static const char *nss_name = "winbind";
 static int nss_errno;
 static NSS_STATUS last_error;
