--- samba-2.2.1a/source/smbd/posix_acls.c~	Fri Jul  6 04:02:02 2001
+++ samba-2.2.1a/source/smbd/posix_acls.c	Wed Oct  3 03:20:47 2001
@@ -418,7 +418,7 @@
 	*pgrp = (gid_t)-1;
 
 	if(security_info_sent == 0) {
-		DEBUG(0,("unpack_nt_owners: no security info sent !\n"));
+		DEBUG(1,("unpack_nt_owners: no security info sent !\n"));
 		return False;
 	}
 
--- samba-2.2.1a/source/smbd/oplock.c~	Fri Jul  6 04:02:01 2001
+++ samba-2.2.1a/source/smbd/oplock.c	Fri Oct  5 21:13:04 2001
@@ -994,7 +994,7 @@
     {
       if(smb_read_error == READ_TIMEOUT)
       {
-        if( DEBUGLVL( 0 ) )
+        if( DEBUGLVL( 1 ) )
         {
           dbgtext( "request_oplock_break: no response received to oplock " );
           dbgtext( "break request to pid %d ", (int)share_entry->pid );
