--- samba-3.0.21a/source/libsmb/ntlmssp.c.old	2005-10-18 04:45:00.000000000 +0200
+++ samba-3.0.21a/source/libsmb/ntlmssp.c	2006-01-22 13:18:02.000000000 +0100
@@ -384,6 +384,11 @@
 
 void ntlmssp_weaken_keys(NTLMSSP_STATE *ntlmssp_state)
 {
+	/* Nothing to weaken.  We certainly don't want to 'extend' the length... */
+	if (ntlmssp_state->session_key.length < 8) {
+		return;
+	}
+
 	/* Key weakening not performed on the master key for NTLM2
 	   and does not occour for NTLM1.  Therefore we only need
 	   to do this for the LM_KEY.
