diff -urN samba-2.2.0.org/source/configure.in samba-2.2.0/source/configure.in
--- samba-2.2.0.org/source/configure.in	Wed Apr 18 01:27:35 2001
+++ samba-2.2.0/source/configure.in	Wed Apr 18 01:29:23 2001
@@ -483,7 +483,7 @@
     AC_CHECK_HEADERS(readline/history.h)
 
     AC_CHECK_HEADERS(readline.h readline/readline.h,[
-      for termlib in ncurses curses termcap terminfo termlib; do
+      for termlib in tinfo ncurses curses termcap terminfo termlib; do
        AC_CHECK_LIB(${termlib}, tgetent, [TERMLIBS="-l${termlib}"; break])
       done
       AC_CHECK_LIB(readline, rl_callback_handler_install,
@@ -510,7 +510,7 @@
     AC_CHECK_HEADERS(readline/history.h)
 
     AC_CHECK_HEADERS(readline.h readline/readline.h,[
-      for termlib in ncurses curses termcap terminfo termlib; do
+      for termlib in tinfo ncurses curses termcap terminfo termlib; do
        AC_CHECK_LIB(${termlib}, tgetent, [TERMLIBS="-l${termlib}"; break])
       done
       AC_CHECK_LIB(readline, rl_callback_handler_install,
