diff -durN samba-3.0.1.orig/source/Makefile.in samba-3.0.1/source/Makefile.in
--- samba-3.0.1.orig/source/Makefile.in	2003-12-10 22:59:18.000000000 +0100
+++ samba-3.0.1/source/Makefile.in	2004-12-30 13:53:05.263072472 +0100
@@ -1243,9 +1243,9 @@
 	@$(SHELL) $(srcdir)/script/installswat.sh $(DESTDIR)$(SWATDIR) $(srcdir)
 
 installclientlib: installdirs libsmbclient
-	@$(SHELL) $(srcdir)/script/installdirs.sh $(DESTDIR)${prefix}/lib
-	-$(INSTALLCLIENTCMD_SH) bin/libsmbclient.@SHLIBEXT@ $(DESTDIR)${prefix}/lib
-	-$(INSTALLCLIENTCMD_A) bin/libsmbclient.a $(DESTDIR)${prefix}/lib
+	@$(SHELL) $(srcdir)/script/installdirs.sh $(DESTDIR)${prefix}/lib64
+	-$(INSTALLCLIENTCMD_SH) bin/libsmbclient.@SHLIBEXT@ $(DESTDIR)${prefix}/lib64
+	-$(INSTALLCLIENTCMD_A) bin/libsmbclient.a $(DESTDIR)${prefix}/lib64
 	@$(SHELL) $(srcdir)/script/installdirs.sh $(DESTDIR)${prefix}/include
 	-$(INSTALLCMD) $(srcdir)/include/libsmbclient.h $(DESTDIR)${prefix}/include
 
diff -durN samba-3.0.1.orig/source/aclocal.m4 samba-3.0.1/source/aclocal.m4
--- samba-3.0.1.orig/source/aclocal.m4	2003-11-07 18:37:32.000000000 +0100
+++ samba-3.0.1/source/aclocal.m4	2004-12-30 13:54:05.767874344 +0100
@@ -496,12 +496,12 @@
   $1=[$]ac_new_flags
 ])
     
-dnl Removes -L/usr/lib/? from given variable
+dnl Removes -L/usr/lib64/? from given variable
 AC_DEFUN(LIB_REMOVE_USR_LIB,[
   ac_new_flags=""
   for i in [$]$1; do
     case [$]i in
-    -L/usr/lib|-L/usr/lib/) ;;
+    -L/usr/lib64|-L/usr/lib64/) ;;
     *) ac_new_flags="[$]ac_new_flags [$]i" ;;
     esac
   done
@@ -632,9 +632,9 @@
 ])
 
 dnl LIB_ADD_DIR(LDFLAGS, $LIBDIR)
-dnl This function doesn't add -L/usr/lib into LDFLAGS
+dnl This function doesn't add -L/usr/lib64 into LDFLAGS
 AC_DEFUN(LIB_ADD_DIR,[
-if test "$2" != "/usr/lib" ; then
+if test "$2" != "/usr/lib64" ; then
     $1="$$1 -L$2"
 fi
 ])
diff -durN samba-3.0.1.orig/source/configure.in samba-3.0.1/source/configure.in
--- samba-3.0.1.orig/source/configure.in	2003-12-15 04:36:25.000000000 +0100
+++ samba-3.0.1/source/configure.in	2004-12-30 13:56:21.871183496 +0100
@@ -22,7 +22,7 @@
     mandir="\${prefix}/share/man"
     logfilebase="\${VARDIR}/log/samba"
     privatedir="\${CONFIGDIR}/private"
-    libdir="\${prefix}/lib/samba"
+    libdir="\${prefix}/lib64/samba"
     swatdir="\${DATADIR}/samba/swat",
     configdir="\${LIBDIR}"
     logfilebase="\${VARDIR}"
@@ -825,7 +825,7 @@
     _cppflags=${CPPFLAGS}
 
     # Add additional search path
-    LDFLAGS="-L$with_readline/lib $LDFLAGS"
+    LDFLAGS="-L$with_readline/lib64 $LDFLAGS"
     CPPFLAGS="-I$with_readline/include $CPPFLAGS"
 
     AC_CHECK_HEADERS(readline.h history.h readline/readline.h)
@@ -836,7 +836,7 @@
        AC_CHECK_LIB(${termlib}, tgetent, [TERMLIBS="-l${termlib}"; break])
       done
       AC_CHECK_LIB(readline, rl_callback_handler_install,
-       [TERMLDFLAGS="-L$with_readline/lib"
+       [TERMLDFLAGS="-L$with_readline/lib64"
        TERMCPPFLAGS="-I$with_readline/include"
        CPPFLAGS="-I$with_readline/include $CPPFLAGS"
        TERMLIBS="-lreadline $TERMLIBS"
@@ -1615,7 +1615,7 @@
 ICONV_LOCATION=standard
 LOOK_DIRS="/usr /usr/local /sw"
 AC_ARG_WITH(libiconv,
-[  --with-libiconv=BASEDIR Use libiconv in BASEDIR/lib and BASEDIR/include (default=auto) ],
+[  --with-libiconv=BASEDIR Use libiconv in BASEDIR/lib64 and BASEDIR/include (default=auto) ],
 [
   if test "$withval" = "no" ; then
     AC_MSG_ERROR([argument to --with-libiconv must be a directory])
@@ -1635,7 +1635,7 @@
     CPPFLAGS="$CPPFLAGS -I$i/include"
 dnl This is here to handle -withval stuff for --with-libiconv
 dnl Perhaps we should always add a -L
-    LDFLAGS="$LDFLAGS -L$i/lib"
+    LDFLAGS="$LDFLAGS -L$i/lib64"
     LIBS=
     export LDFLAGS LIBS CPPFLAGS
 dnl Try to find iconv(3)
@@ -1643,7 +1643,7 @@
 
     if test "$ICONV_FOUND" = yes; then
 	LDFLAGS=$save_LDFLAG
-        LIB_ADD_DIR(LDFLAGS, "$i/lib")
+        LIB_ADD_DIR(LDFLAGS, "$i/lib64")
         CFLAGS_ADD_DIR(CPPFLAGS, "$i/include")
         LIBS="$save_LIBS"
         ICONV_LOCATION=$i
@@ -1655,7 +1655,7 @@
         # check for iconv in libc
         ic_save_LIBS="$LIBS"
         if test x"$ICONV_PATH_SPEC" = "xyes" ; then
-           LIBS="$LIBS -L$ICONV_LOCATION/lib"
+           LIBS="$LIBS -L$ICONV_LOCATION/lib64"
         fi
         if test x"$jm_cv_lib_iconv" != x; then
            LIBS="$LIBS -l$jm_cv_lib_iconv"
@@ -1719,7 +1719,7 @@
            fi
 	   dnl Add the flags we need to CPPFLAGS and LDFLAGS
 	   CFLAGS_ADD_DIR(CPPFLAGS, "$i/include")
-	   LIB_ADD_DIR(LDFLAGS, "$i/lib")
+	   LIB_ADD_DIR(LDFLAGS, "$i/lib64")
 	   export CPPFLAGS
            AC_DEFINE(HAVE_NATIVE_ICONV,1,[Whether to use native iconv])
 	   AC_DEFINE_UNQUOTED(DEFAULT_DOS_CHARSET,$default_dos_charset,[Default dos charset name])
@@ -2468,7 +2468,7 @@
         AC_MSG_RESULT($withval)
         KRB5_CFLAGS="-I$withval/include"
         KRB5_CPPFLAGS="-I$withval/include"
-        KRB5_LDFLAGS="-L$withval/lib"
+        KRB5_LDFLAGS="-L$withval/lib64"
         FOUND_KRB5=yes
         ;;
       esac ],
@@ -2481,10 +2481,10 @@
     # see if this box has the SuSE location for the heimdal krb implementation
     AC_MSG_CHECKING(for /usr/include/heimdal)
     if test -d /usr/include/heimdal; then
-      if test -f /usr/lib/heimdal/lib/libkrb5.a; then
+      if test -f /usr/lib64/heimdal/lib/libkrb5.a; then
           KRB5_CFLAGS="-I/usr/include/heimdal"
           KRB5_CPPFLAGS="-I/usr/include/heimdal"
-          KRB5_LDFLAGS="-L/usr/lib/heimdal/lib"
+          KRB5_LDFLAGS="-L/usr/lib64/heimdal/lib"
           AC_MSG_RESULT(yes)
       else
           KRB5_CFLAGS="-I/usr/include/heimdal"
@@ -4261,7 +4261,7 @@
 builddir=`pwd`
 AC_SUBST(builddir)
 
-dnl Remove -L/usr/lib/? from LDFLAGS and LIBS
+dnl Remove -L/usr/lib64/? from LDFLAGS and LIBS
 LIB_REMOVE_USR_LIB(LDFLAGS)
 LIB_REMOVE_USR_LIB(LIBS)
 
diff -durN samba-3.0.1.orig/source/torture/nsstest.c samba-3.0.1/source/torture/nsstest.c
--- samba-3.0.1.orig/source/torture/nsstest.c	2003-08-15 22:40:00.000000000 +0200
+++ samba-3.0.1/source/torture/nsstest.c	2004-12-30 13:57:10.644768784 +0100
@@ -21,7 +21,7 @@
 
 #include "includes.h"
 
-static const char *so_path = "/lib/libnss_winbind.so";
+static const char *so_path = "/lib64/libnss_winbind.so";
 static const char *nss_name = "winbind";
 static int nss_errno;
 static NSS_STATUS last_error;
