--- samba-3.4.5/source4/torture/rpc/spoolss.c~	2010-01-18 12:38:09.000000000 +0100
+++ samba-3.4.5/source4/torture/rpc/spoolss.c	2010-01-26 17:03:53.363835346 +0100
@@ -2087,6 +2087,7 @@
 	return true;
 }
 
+#if 0
 bool test_printer_keys(struct torture_context *tctx,
 		       struct dcerpc_pipe *p,
 		       struct policy_handle *handle)
@@ -2144,6 +2145,7 @@
 
 	return true;
 }
+#endif
 
 bool torture_rpc_spoolss(struct torture_context *torture)
 {
