--- samba-3.5.2/source3/param/loadparm.c.orig	2010-05-04 20:18:14.000000000 +0200
+++ samba-3.5.2/source3/param/loadparm.c	2010-05-04 20:38:34.882642212 +0200
@@ -5093,7 +5093,7 @@
 	   a large number of sites (tridge) */
 	Globals.bHostnameLookups = False;
 
-	string_set(&Globals.szPassdbBackend, "tdbsam");
+	string_set(&Globals.szPassdbBackend, "smbpasswd");
 	string_set(&Globals.szLdapSuffix, "");
 	string_set(&Globals.szLdapMachineSuffix, "");
 	string_set(&Globals.szLdapUserSuffix, "");
