--- samba-3.0.21/source3/include/includes.h.orig	2005-12-25 18:08:07.204427000 +0100
+++ samba-3.0.21/source3/include/includes.h	2005-12-25 18:11:14.004427000 +0100
@@ -33,7 +33,7 @@
 #define protected #error DONT_USE_CPLUSPLUS_RESERVED_NAMES
 #define template #error DONT_USE_CPLUSPLUS_RESERVED_NAMES
 #define this #error DONT_USE_CPLUSPLUS_RESERVED_NAMES
-#define new #error DONT_USE_CPLUSPLUS_RESERVED_NAMES
+/* #define new #error DONT_USE_CPLUSPLUS_RESERVED_NAMES */
 #define delete #error DONT_USE_CPLUSPLUS_RESERVED_NAMES
 #define friend #error DONT_USE_CPLUSPLUS_RESERVED_NAMES
 #endif
