--- samba-3.0.21/source/python/setup.py.orig	2005-12-25 20:27:49.884427000 +0100
+++ samba-3.0.21/source/python/setup.py	2005-12-25 20:32:21.984427000 +0100
@@ -63,8 +63,10 @@
         next_is_flag = 0;
     elif lib == "-Wl,-rpath":
         next_is_path = 1;
-    elif lib[0:2] in ("-l","-pthread"):
+    elif lib[0:2] == "-l":
         libraries.append(lib[2:])
+    elif lib[0:8] == "-pthread":
+	libraries.append(lib[1:])
     elif lib[0:2] == "-L":
         library_dirs.append(lib[2:])
     elif lib[0:2] in ("-W","-s"):
