diff -urN samba-2.2.0.org/source/Makefile.in samba-2.2.0/source/Makefile.in
--- samba-2.2.0.org/source/Makefile.in	Wed Apr 18 01:27:35 2001
+++ samba-2.2.0/source/Makefile.in	Wed Apr 18 01:38:13 2001
@@ -30,15 +30,15 @@
 top_builddir=.
 SHELL=/bin/sh
 
-BASEDIR= @prefix@
-BINDIR = @bindir@
-DATADIR = @datadir@
+BASEDIR= $(DESTDIR)@prefix@
+BINDIR = $(DESTDIR)@bindir@
+DATADIR = $(DESTDIR)@datadir@
 # sbindir is mapped to bindir when compiling SAMBA in 2.0.x compatibility mode.
-SBINDIR = @sbindir@
-LIBDIR = @libdir@
-VARDIR = @localstatedir@
-CONFIGDIR = @configdir@
-MANDIR = @mandir@
+SBINDIR = $(DESTDIR)@sbindir@
+LIBDIR = $(DESTDIR)@libdir@
+VARDIR = $(DESTDIR)@localstatedir@
+CONFIGDIR = $(DESTDIR)@configdir@
+MANDIR = $(DESTDIR)@mandir@
 
 # The permissions to give the executables
 INSTALLPERMS = 0755
@@ -58,7 +58,7 @@
 TDB_PASSWD_FILE = $(PRIVATEDIR)/smbpasswd.tdb
 
 # This is where SWAT images and help files go
-SWATDIR = @swatdir@
+SWATDIR = $(DESTDIR)@swatdir@
 
 # the directory where lock files go
 LOCKDIR = @lockdir@
diff -urN samba-2.2.0.org/source/script/installbin.sh samba-2.2.0/source/script/installbin.sh
--- samba-2.2.0.org/source/script/installbin.sh	Wed Apr 18 01:27:39 2001
+++ samba-2.2.0/source/script/installbin.sh	Wed Apr 18 01:36:43 2001
@@ -13,7 +13,7 @@
 
 for d in $BASEDIR $BINDIR $LIBDIR $VARDIR $BASEDIR/private; do
 if [ ! -d $d ]; then
-mkdir $d
+mkdir -p $d
 if [ ! -d $d ]; then
   echo Failed to make directory $d
   exit 1
@@ -34,7 +34,7 @@
 
  # this is a special case, mount needs this in a specific location
  if [ $p2 = smbmount ]; then
-   ln -sf $BINDIR/$p2 /sbin/mount.smbfs
+   ln -sf /usr/sbin/$p2 $DESTDIR/$SBINDIR/mount.smbfs
  fi
 done
 
diff -urN samba-2.2.0.org/source/script/installcp.sh samba-2.2.0/source/script/installcp.sh
--- samba-2.2.0.org/source/script/installcp.sh	Wed Apr 18 01:27:39 2001
+++ samba-2.2.0/source/script/installcp.sh	Wed Apr 18 01:36:43 2001
@@ -12,7 +12,7 @@
 echo Installing codepage files in $CODEPAGEDIR
 for d in $LIBDIR $CODEPAGEDIR; do
 if [ ! -d $d ]; then
-mkdir $d
+mkdir -p $d
 if [ ! -d $d ]; then
   echo Failed to make directory $d
   exit 1
@@ -23,11 +23,11 @@
 for p in $*; do
  if [ -f ${srcdir}/codepages/codepage_def.$p ]; then
    echo Creating codepage file $CODEPAGEDIR/codepage.$p
-   $BINDIR/make_smbcodepage c $p ${srcdir}/codepages/codepage_def.$p $CODEPAGEDIR/codepage.$p
+   ${srcdir}/bin/make_smbcodepage c $p ${srcdir}/codepages/codepage_def.$p $CODEPAGEDIR/codepage.$p
  fi
  if [ -f ${srcdir}/codepages/CP${p}.TXT ]; then
    echo Creating unicode map $CODEPAGEDIR/unicode_map.$p
-   $BINDIR/make_unicodemap $p ${srcdir}/codepages/CP${p}.TXT $CODEPAGEDIR/unicode_map.$p
+   ${srcdir}/bin/make_unicodemap $p ${srcdir}/codepages/CP${p}.TXT $CODEPAGEDIR/unicode_map.$p
  fi
 done
 
diff -urN samba-2.2.0.org/source/script/installman.sh samba-2.2.0/source/script/installman.sh
--- samba-2.2.0.org/source/script/installman.sh	Wed Apr 18 01:27:39 2001
+++ samba-2.2.0/source/script/installman.sh	Wed Apr 18 01:36:43 2001
@@ -11,7 +11,7 @@
 
 for d in $MANDIR $MANDIR/man1 $MANDIR/man5 $MANDIR/man7 $MANDIR/man8; do
 if [ ! -d $d ]; then
-mkdir $d
+mkdir -p $d
 if [ ! -d $d ]; then
   echo Failed to make directory $d, does $USER have privileges?
   exit 1
diff -urN samba-2.2.0.org/source/script/installscripts.sh samba-2.2.0/source/script/installscripts.sh
--- samba-2.2.0.org/source/script/installscripts.sh	Wed Apr 18 01:27:39 2001
+++ samba-2.2.0/source/script/installscripts.sh	Wed Apr 18 01:36:43 2001
@@ -12,7 +12,7 @@
 
 for d in $BINDIR; do
  if [ ! -d $d ]; then
-  mkdir $d
+  mkdir -p $d
   if [ ! -d $d ]; then
     echo Failed to make directory $d
     echo Have you run installbin first?
diff -urN samba-2.2.0.org/source/script/installswat.sh samba-2.2.0/source/script/installswat.sh
--- samba-2.2.0.org/source/script/installswat.sh	Wed Apr 18 01:27:39 2001
+++ samba-2.2.0/source/script/installswat.sh	Wed Apr 18 01:38:55 2001
@@ -10,7 +10,7 @@
 
 for d in $SWATDIR $SWATDIR/help $SWATDIR/images $SWATDIR/include; do
     if [ ! -d $d ]; then
-	mkdir $d
+	mkdir -p $d
 	if [ ! -d $d ]; then
 	    echo Failed to make directory $d, does $USER have privileges?
 	    exit 1
@@ -69,7 +69,7 @@
 
     for d in $BOOKDIR $BOOKDIR/figs $BOOKDIR/gifs; do
         if [ ! -d $d ]; then
-            mkdir $d
+            mkdir -p $d
             if [ ! -d $d ]; then
                 echo Failed to make directory $d, does $USER have privileges?
                 exit 1
