diff -urN samba-3.0.14a.org/source/python/setup.py samba-3.0.14a/source/python/setup.py
--- samba-3.0.14a.org/source/python/setup.py	2005-02-25 18:59:23.000000000 +0100
+++ samba-3.0.14a/source/python/setup.py	2005-07-08 22:51:27.823948560 +0200
@@ -49,8 +49,12 @@
 
 libraries = []
 library_dirs = []
+skip_next_keyword = 0
 
 for lib in string.split(samba_libs):
+    if (skip_next_keyword):
+	skip_next_keyword = 0
+	continue
     if lib[0:2] == "-l":
         libraries.append(lib[2:])
         continue
@@ -60,6 +64,12 @@
     if lib[0:2] == "-W":
         # Skip linker flags
         continue
+    if lib[0:2] == "-z":
+	skip_next_keyword = 1
+	continue
+    if lib[0:8] == "-pthread":
+	library_dirs.append("-lpthread")
+	continue
     print "Unknown entry '%s' in $LIBS variable passed to setup.py" % lib
     sys.exit(1)
 
