diff -urN samba-3.0.25a.org/examples/libsmbclient/smbwrapper/Makefile samba-3.0.25a/examples/libsmbclient/smbwrapper/Makefile
--- samba-3.0.25a.org/examples/libsmbclient/smbwrapper/Makefile	2006-01-25 00:46:42.000000000 +0100
+++ samba-3.0.25a/examples/libsmbclient/smbwrapper/Makefile	2007-06-17 17:43:30.088457851 +0200
@@ -3,7 +3,7 @@
 
 CFLAGS = -I$(SAMBA_INCL) $(EXTLIB_INCL)
 
-LDFLAGS = -L/usr/local/samba/lib
+LDFLAGS = -L../../../source/bin
 
 SMBINCLUDE = -I../../../source/include
 CFLAGS= -fpic -g -O0 $(DEFS) $(SMBINCLUDE)
diff -urN samba-3.0.25a.org/examples/libsmbclient/smbwrapper/wrapper.c samba-3.0.25a/examples/libsmbclient/smbwrapper/wrapper.c
--- samba-3.0.25a.org/examples/libsmbclient/smbwrapper/wrapper.c	2006-01-25 00:46:42.000000000 +0100
+++ samba-3.0.25a/examples/libsmbclient/smbwrapper/wrapper.c	2007-06-17 17:39:16.085693163 +0200
@@ -62,6 +62,7 @@
 #include <dirent.h>
 #include <signal.h>
 #include <stdarg.h>
+#include <string.h>
 #ifdef __USE_GNU
 # define SMBW_USE_GNU
 #endif
@@ -73,6 +74,7 @@
 #include <errno.h>
 #include "libsmbclient.h"
 #include "bsd-strlfunc.h"
+#include "smbw.h"
 #include "wrapper.h"
 
 /*
@@ -488,7 +490,7 @@
          * function and smbw.c et al.  Functions that return other than an
          * integer value, however, MUST have their return value defined.
          */
-        off64_t         smbw_lseek();
+        /* off64_t         smbw_lseek(); */
         
         if (smbw_fd(fd)) {
 		return (off_t) smbw_lseek(fd, offset, whence);
@@ -518,7 +520,7 @@
          * function and smbw.c et al.  Functions that return other than an
          * integer value, however, MUST have their return value defined.
          */
-        off64_t         smbw_lseek();
+        /* off64_t         smbw_lseek(); */
         
 	if (smbw_fd(fd))
 		ret = smbw_lseek(fd, offset, whence);
