commit 1f44b4aaa5f700827ee2ab272ae4b59e559b094b
Author: Günther Deschner <gd@samba.org>
Date:   Fri Sep 5 14:01:45 2008 +0200

    build: don't install the cifs.upcall binary twice.
    
    Guenther

diff --git a/source/Makefile.in b/source/Makefile.in
index 8b46ab6..b67cfe8 100644
--- a/source/Makefile.in
+++ b/source/Makefile.in
@@ -178,7 +178,7 @@ PATH_FLAGS = -DSMB_PASSWD_FILE=\"$(SMB_PASSWD_FILE)\" \
 
 # Note that all executable programs now provide for an optional executable suffix.
 
-SBIN_PROGS = bin/smbd@EXEEXT@ bin/nmbd@EXEEXT@ @SWAT_SBIN_TARGETS@ @EXTRA_SBIN_PROGS@ @CIFSUPCALL_PROGS@
+SBIN_PROGS = bin/smbd@EXEEXT@ bin/nmbd@EXEEXT@ @SWAT_SBIN_TARGETS@ @EXTRA_SBIN_PROGS@
 
 ROOT_SBIN_PROGS = @CIFSMOUNT_PROGS@
 
