--- samba-2.2.8a/source/libsmb/clientgen.c~	2002-12-10 15:58:15.000000000 +0100
+++ samba-2.2.8a/source/libsmb/clientgen.c	2003-09-11 19:00:11.000000000 +0200
@@ -170,7 +170,9 @@
 	/* Check the effective uid - make sure we are not setuid */
 	if (is_setuid_root()) {
 		DEBUG(0,("libsmb based programs must *NOT* be setuid root.\n"));
+		/* Warn but don't exit, amanda must be setuid and needs smb
 		return NULL;
+		*/
 	}
 
 	if (!cli) {
