--- samba-3.0.23b.org/source/torture/nsstest.c	2006-04-20 04:29:44.000000000 +0200
+++ samba-3.0.23b/source/torture/nsstest.c	2006-08-09 21:19:38.922291250 +0200
@@ -21,7 +21,7 @@
 
 #include "includes.h"
 
-static const char *so_path = "/lib/libnss_winbind.so";
+static const char *so_path = "/lib64/libnss_winbind.so";
 static const char *nss_name = "winbind";
 static int nss_errno;
 static NSS_STATUS last_error;
--- samba-3.0.25a/source/aclocal.m4.orig	2007-03-01 05:55:17.000000000 +0100
+++ samba-3.0.25a/source/aclocal.m4	2007-06-14 18:45:50.703496205 +0200
@@ -268,20 +268,20 @@
   $1=[$]ac_new_flags
 ])
 
-dnl Removes '-L/usr/lib[/]', '-Wl,-rpath,/usr/lib[/]'
-dnl and '-Wl,-rpath -Wl,/usr/lib[/]' from given variable
+dnl Removes '-L/usr/lib64[/]', '-Wl,-rpath,/usr/lib64[/]'
+dnl and '-Wl,-rpath -Wl,/usr/lib64[/]' from given variable
 AC_DEFUN(LIB_REMOVE_USR_LIB,[
   ac_new_flags=""
   l=""
   for i in [$]$1; do
     case [$]l[$]i in
-    -L/usr/lib) ;;
-    -L/usr/lib/) ;;
-    -Wl,-rpath,/usr/lib) ;;
-    -Wl,-rpath,/usr/lib/) ;;
+    -L/usr/lib64) ;;
+    -L/usr/lib64/) ;;
+    -Wl,-rpath,/usr/lib64) ;;
+    -Wl,-rpath,/usr/lib64/) ;;
     -Wl,-rpath) l=[$]i;;
-    -Wl,-rpath-Wl,/usr/lib) l="";;
-    -Wl,-rpath-Wl,/usr/lib/) l="";;
+    -Wl,-rpath-Wl,/usr/lib64) l="";;
+    -Wl,-rpath-Wl,/usr/lib64/) l="";;
     *)
     	s=" "
         if test x"[$]ac_new_flags" = x""; then
@@ -429,9 +429,9 @@
 ])
 
 dnl LIB_ADD_DIR(LDFLAGS, $LIBDIR)
-dnl This function doesn't add -L/usr/lib into LDFLAGS
+dnl This function doesn't add -L/usr/lib64 into LDFLAGS
 AC_DEFUN(LIB_ADD_DIR,[
-if test "$2" != "/usr/lib" ; then
+if test "$2" != "/usr/lib64" ; then
     $1="$$1 -L$2"
 fi
 ])
--- samba-3.0.25a/source/configure.in.orig	2007-05-23 17:29:20.000000000 +0200
+++ samba-3.0.25a/source/configure.in	2007-06-14 18:49:28.307646759 +0200
@@ -31,7 +31,7 @@
 test "${mandir}" || mandir="\${prefix}/man"
 logfilebase="\${VARDIR}"
 privatedir="\${prefix}/private"
-test "${libdir}" || libdir="\${prefix}/lib"
+test "${libdir}" || libdir="\${prefix}/lib64"
 pammodulesdir="\${LIBDIR}/security"
 configdir="\${LIBDIR}"
 swatdir="\${prefix}/swat"
@@ -45,7 +35,7 @@
     mandir="\${prefix}/share/man"
     logfilebase="\${VARDIR}/log/samba"
     privatedir="\${CONFIGDIR}/private"
-    libdir="\${prefix}/lib/samba"
+    libdir="\${prefix}/lib64/samba"
     configdir="\${sysconfdir}/samba"
     swatdir="\${DATADIR}/samba/swat"
     ;;
@@ -1132,7 +1132,7 @@
     _cppflags=${CPPFLAGS}
 
     # Add additional search path
-    LDFLAGS="-L$with_readline/lib $LDFLAGS"
+    LDFLAGS="-L$with_readline/lib64 $LDFLAGS"
     CPPFLAGS="-I$with_readline/include $CPPFLAGS"
 
     AC_CHECK_HEADERS(readline.h history.h readline/readline.h)
@@ -1143,7 +1143,7 @@
        AC_CHECK_LIB(${termlib}, tgetent, [TERMLIBS="-l${termlib}"; break])
       done
       AC_CHECK_LIB(readline, rl_callback_handler_install,
-       [TERMLDFLAGS="-L$with_readline/lib"
+       [TERMLDFLAGS="-L$with_readline/lib64"
        TERMCPPFLAGS="-I$with_readline/include"
        CPPFLAGS="-I$with_readline/include $CPPFLAGS"
        TERMLIBS="-lreadline $TERMLIBS"
@@ -2337,7 +2337,7 @@
 ICONV_LOCATION=standard
 LOOK_DIRS="/usr /usr/local /sw /opt"
 AC_ARG_WITH(libiconv,
-[  --with-libiconv=BASEDIR Use libiconv in BASEDIR/lib and BASEDIR/include (default=auto) ],
+[  --with-libiconv=BASEDIR Use libiconv in BASEDIR/lib64 and BASEDIR/include (default=auto) ],
 [
   if test "$withval" = "no" ; then
     AC_MSG_ERROR([argument to --with-libiconv must be a directory])
@@ -3370,7 +3370,7 @@
         AC_MSG_RESULT($withval)
         KRB5_CFLAGS="-I$withval/include"
         KRB5_CPPFLAGS="-I$withval/include"
-        KRB5_LDFLAGS="-L$withval/lib"
+        KRB5_LDFLAGS="-L$withval/lib64"
         FOUND_KRB5=yes
 	if test -x "$withval/bin/krb5-config"; then
 		KRB5CONFIG=$withval/bin/krb5-config
@@ -3407,10 +3407,10 @@
     # see if this box has the SuSE location for the heimdal krb implementation
     AC_MSG_CHECKING(for /usr/include/heimdal)
     if test -d /usr/include/heimdal; then
-      if test -f /usr/lib/heimdal/lib/libkrb5.a; then
+      if test -f /usr/lib64/heimdal/lib/libkrb5.a; then
           KRB5_CFLAGS="-I/usr/include/heimdal"
           KRB5_CPPFLAGS="-I/usr/include/heimdal"
-          KRB5_LDFLAGS="-L/usr/lib/heimdal/lib"
+          KRB5_LDFLAGS="-L/usr/lib64/heimdal/lib"
           AC_MSG_RESULT(yes)
       else
           KRB5_CFLAGS="-I/usr/include/heimdal"
@@ -6175,7 +6175,7 @@
 
 fi
 
-dnl Remove -L/usr/lib/? from LDFLAGS and LIBS
+dnl Remove -L/usr/lib64/? from LDFLAGS and LIBS
 LIB_REMOVE_USR_LIB(LDFLAGS)
 LIB_REMOVE_USR_LIB(LIBS)
 LIB_REMOVE_USR_LIB(KRB5_LIBS)
